docs/branch-operations.md

```diff
@@ -0,0 +1,540 @@
+1. Collects all flakes (assertions/retractions) from source commits
+2. Groups them by `[subject predicate object datatype meta]`
+3. Counts net operations (assertions - retractions) for each unique value
+4. Produces final flakes:
+   - Net positive → assertion
+   - Net negative → retraction  
+   - Net zero → cancelled out (no flake)
+
+This means that if you assert a value in one commit and retract it in another, they cancel out and disappear from the squashed result.
+
+Future support planned for:
+- Regular 3-way merge (without squash)
+- Schema-aware conflict resolution when validation is added
+
+#### Options
+
+| Option | Type | Default | Description |
+|--------|------|---------|-------------|
+| `ff` | keyword | `:auto` | Fast-forward behavior: `:auto`, `:only`, `:never` |
+| `squash?` | boolean | `false` | Combine all commits into one |
+| `message` | string | auto-generated | Commit message for the new commit |
+| `preview?` | boolean | `false` | Dry run without making changes |
+
+**Note on `:message`:**
+- The `:message` option sets the commit message for the new commit created by the merge
+- If not provided, auto-generates: `"Squash merge from <source-branch>"` or similar
+- This is stored as the commit's message property (separate from annotation)
+- For fast-forward operations, no new commit is created so the message is not used
+
+#### Examples
+
+##### Standard Merge (Auto fast-forward when possible)
+```clojure
+;; Automatically fast-forwards if possible, otherwise fails
+(fluree/merge! conn "ledger:feature" "ledger:main")
+```
+
+##### Squash Merge
+```clojure
+;; Combines all commits from feature branch into single commit on main
+(fluree/merge! conn "ledger:feature" "ledger:main"
+  {:squash? true
+   :message "Feature X implementation"})
+```
+
+##### Fast-forward Only
+```clojure
+(fluree/merge! conn "ledger:feature" "ledger:main"
+  {:ff :only})  ; Fail if fast-forward not possible
+```
+
+#### Response
+
+```clojure
+{:status :success  ; or :error
+ :operation :merge
+ :from "ledger:feature"
+ :to "ledger:main" 
+ :strategy "fast-forward"  ; or "squash"
+ :commits {:merged 3}  ; number of commits merged
+ :new-commit "sha-of-new-commit"}  ; for squash merge
+```
+
+#### Implementation Status
+**Currently Implemented:**
+- Fast-forward (`:ff :auto`, `:ff :only`, `:ff :never`)
+- Squash (`:squash? true`)
+
+**Not Yet Implemented:**
+- Regular 3-way merge (without squash)
+- Conflict detection and resolution (will come with SHACL/policy validation)
+
+---
+
+### 3. RESET - Safe Rollback or Hard Reset
+
+Resets a branch to a previous state.
+
+#### Implementation Status
+
+| Mode | Status | Description |
+|------|--------|-------------|
+| Safe mode (`:mode :safe`) | ✅ Implemented | Creates a revert commit to target state |
+| Hard mode (`:mode :hard`) | ❌ Not Implemented | Will move branch pointer (rewrite history) |
+| Preview (`:preview? true`) | ✅ Implemented | Dry-run to see what would happen |
```

Comment:
Just a thought, but if we're copying `git`'s api we might want to call this `dry-run?` instead : )